--- gcc-6.2.0/config.sub	2015-12-31 21:13:28.000000000 +0000
+++ gcc-6.2.0-shk/config.sub	2016-09-18 19:01:25.702252200 +0100
@@ -1378,7 +1378,7 @@
 	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
 	      | -sym* | -kopensolaris* | -plan9* \
 	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
-	      | -aos* | -aros* | -cloudabi* | -sortix* \
+	      | -aos* | -aros* | -cloudabi* | -sortix* | -shk* \
 	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
 	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
 	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
--- gcc-6.2.0/gcc/config.gcc	2016-06-08 14:34:25.000000000 +0100
+++ gcc-6.2.0-shk/gcc/config.gcc	2016-09-18 21:11:40.835949600 +0100
@@ -612,6 +612,11 @@
 
 # Common parts for widely ported systems.
 case ${target} in
+*-*-shk*)
+  gas=yes
+  gnu_ld=yes
+  default_use_cxa_atexit=yes
+  ;;
 *-*-darwin*)
   tmake_file="t-darwin ${cpu_type}/t-darwin"
   tm_file="${tm_file} darwin.h"
@@ -907,6 +912,9 @@
 esac
 
 case ${target} in
+i386-*-shk*)
+	tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h glibc-stdint.h i386/i386elf.h shk.h"
+	;;
 aarch64*-*-elf | aarch64*-*-rtems*)
 	tm_file="${tm_file} dbxelf.h elfos.h newlib-stdint.h"
 	tm_file="${tm_file} aarch64/aarch64-elf.h aarch64/aarch64-elf-raw.h"
--- gcc-6.2.0/gcc/config/shk.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-6.2.0-shk/gcc/config/shk.h	2016-09-18 22:36:31.326167400 +0100
@@ -0,0 +1,29 @@
+/* Useful if you wish to make target-specific gcc changes. */
+#undef TARGET_SHK
+#define TARGET_SHK 1
+
+/* Default arguments you want when running your
+   i386-shk-gcc toolchain */
+#define LIB_SPEC "-lc -lg -lm -lnosys" /* link against C standard library */
+
+/* Files that are linked before user code.
+   The %s tells gcc to look for these files in the library directory. */
+#define STARTFILE_SPEC "crt0.o%s crti.o%s crtbegin.o%s"
+
+/* Files that are linked after user code. */
+#define ENDFILE_SPEC "crtend.o%s crtn.o%s"
+
+/* Don't automatically add extern "C" { } around header files. */
+#undef  NO_IMPLICIT_EXTERN_C
+#define NO_IMPLICIT_EXTERN_C 1
+
+/* Additional predefined macros. */
+#undef TARGET_OS_CPP_BUILTINS
+#define TARGET_OS_CPP_BUILTINS()      \
+  do {                                \
+    builtin_define ("__shk__");      \
+    builtin_define ("__unix__");      \
+    builtin_assert ("system=shk");   \
+    builtin_assert ("system=unix");   \
+    builtin_assert ("system=posix");  \
+  } while(0);
--- gcc-6.2.0/libstdc++-v3/crossconfig.m4	2015-10-11 23:34:28.000000000 +0100
+++ gcc-6.2.0-shk/libstdc++-v3/crossconfig.m4	2016-09-18 19:31:43.752971800 +0100
@@ -5,6 +5,13 @@
 AC_DEFUN([GLIBCXX_CROSSCONFIG],[
 # Base decisions on target environment.
 case "${host}" in
+  *-shk*)
+    GLIBCXX_CHECK_COMPILER_FEATURES
+    GLIBCXX_CHECK_LINKER_FEATURES
+    GLIBCXX_CHECK_MATH_SUPPORT
+    GLIBCXX_CHECK_STDLIB_SUPPORT
+    ;;
+
   arm*-*-symbianelf*)
     # This is a freestanding configuration; there is nothing to do here.
     ;;
--- gcc-6.2.0/libgcc/config.host	2016-05-17 07:15:52.000000000 +0100
+++ gcc-6.2.0-shk/libgcc/config.host	2016-09-18 19:35:23.548344900 +0100
@@ -327,6 +327,10 @@
 esac
 
 case ${host} in
+i386-*-shk*)
+	extra_parts="$extra_parts crtbegin.o crtend.o"
+	tmake_file="${tmake_file} i386/t-crtstuff t-crtstuff-pic t-libgcc-pic"
+	;;
 aarch64*-*-elf | aarch64*-*-rtems*)
 	extra_parts="$extra_parts crtbegin.o crtend.o crti.o crtn.o"
 	extra_parts="$extra_parts crtfastmath.o"
--- gcc-6.2.0/fixincludes/mkfixinc.sh	2015-04-22 15:18:16.000000000 +0100
+++ gcc-6.2.0-shk/fixincludes/mkfixinc.sh	2016-09-18 21:48:55.875726300 +0100
@@ -11,6 +11,7 @@
 
 # Check for special fix rules for particular targets
 case $machine in
+    *-shk* | \
     i?86-*-cygwin* | \
     i?86-*-mingw32* | \
     x86_64-*-mingw32* | \
